extmod/machine_i2c.c

```diff
@@ -172,33 +154,26 @@ STATIC int mp_hal_i2c_read_byte(machine_i2c_obj_t *self, uint8_t *val, int nack)
     return 1; // success
 }
 
-STATIC void mp_hal_i2c_read(machine_i2c_obj_t *self, uint8_t addr, uint8_t *data, size_t len) {
-    mp_hal_i2c_start(self);
-    if (!mp_hal_i2c_write_byte(self, (addr << 1) | 1)) {
-        goto er;
+STATIC int mp_hal_i2c_write_addresses(machine_i2c_obj_t *self, uint8_t addr,
+        uint32_t memaddr, uint8_t addrsize) {
+    if (!mp_hal_i2c_write_byte(self, addr << 1)) {
+        return 0; // error
     }
-    while (len--) {
-        if (!mp_hal_i2c_read_byte(self, data++, len == 0)) {
-            goto er;
+    for (int16_t i = addrsize; i > 0; i -= 8) {
+        if (!mp_hal_i2c_write_byte(self, (memaddr >> (i - 8)) & 0xff)) {
```

Comment:
If addrsize < 8 then funny things will happen with the `>>`.  You can probably make this loop simpler by doing something like `for (int i = addrsize >> 3; i-- > 0;) { if (!mp_hal_i2c_write_byte(self, memaddr >> i)) { ...`.  Note that you don't need to mask `&0xff` because write_byte can't send more than 8 bits.